src/sonic-ctrmgrd/ctrmgr/container_startup.py

```diff
@@ -0,0 +1,293 @@
+    output = proc.communicate()[0].decode("utf-8")
+    return output.strip()[:12]
+
+
+def instance_higher(feature, version):
+    # Compares given version against current version in STATE-DB.
+    # Return True if this version is higher than current.
+    #
+    ret = False
+    ct_version = state_data[VERSION]
+    if ct_version:
+        ct = ct_version.split('.')
+        nxt = version.split('.')
+        for cs, ns in zip(ct, nxt):
+            c = int(cs)
+            n = int(ns)
+            if n < c:
+                break
+            elif n > c:
+                ret = True
+                break
+    else:
+        # Empty version implies no one is running.
+        ret = True
+
+    debug_msg("compare version: new:{} current:{} res={}".format(
+        version, ct_version, ret))
+    return ret
+
+
+def is_active(feature):
+    # Check current system state of the feature
+    if state_data[SYSTEM_STATE] == "up":
+        return True
+    else:
+        syslog.syslog(syslog.LOG_ERR, "Found inactive for {}".format(feature))
+        return False
+
+
+def update_state(feature, owner=None, version=None):
+    """
+    sets owner, version & container-id for this feature in state-db.
+
+    If owner is local, update label to block remote deploying same version or
+    if kube, sets state to "running". 
+
+    """
+    data = {
+            CURRENT_OWNER: owner,
+            DOCKER_ID: get_docker_id() if owner != "local" else feature,
+            UPD_TIMESTAMP: str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")),
+            "hello": "world",
+            VERSION: version
+            }
+
+    if (owner == "local"):
+        # Disable deployment of this version as available locally
+        drop_label(feature, version)
+    else:
+        data[REMOTE_STATE] = "running"
+
+    debug_msg("{} up data:{}".format(feature, str(data)))
+    update_data(feature,  data)
+    state_data.update(data)
+
+
+def do_exit(feat, m):
+    # Exiting will kick off the container to run.
+    # So sleep forever with periodic logs.
+    #
+    while True:
+        syslog.syslog(syslog.LOG_ERR, "Blocking .... feat:{} docker_id:{} msg:{}".format(
+            feat, get_docker_id(), m))
+        if UNIT_TESTING:
+            break
+        time.sleep(60)
+    
+
+def container_up(feature, owner, version):
+    """
+    This is called by container upon post start.
+
+    The container will run its application, only upon this call
+    complete.
+
+    This call does the basic check for if this starting-container can be allowed
+    to run based on current state, and owner & version of this starting
+    container. 
+
+    If allowed to proceed, this info is recorded in state-db and return
+    to enable container start the main application. Else it proceeds to
+    sleep forever, blocking the container from starting the main application.
+
+    """
+    debug_msg("BEGIN")
+    read_data(feature)
```

Comment:
usually, it is better not to use global varilable, read_data can have return value (set_owner, ... etc on) then, we will know what is exactly returned by read_data

Reply:
Agree. It was a one shot synchronous API call *only* and in case it freezes, it does not read any data, so I went slack. But to go by rules, I took them off.